pkg/cache/application.go

```diff
@@ -203,6 +210,48 @@ func (app *Application) getTasks(state string) []*Task {
 	return taskList
 }
 
+// This is called in every scheduling interval,
+// we are not using dispatcher here because we want to
+// make state transition in sync mode in order to prevent
+// generating too many duplicate events. However, it must
+// ensure non of these calls is expensive, usually, they
+// do nothing more than just triggering the state transition.
+func (app *Application) Schedule() {
+	var states = events.States().Application
+	switch app.GetApplicationState() {
+	case states.New:
+		ev := NewSubmitApplicationEvent(app.GetApplicationId())
+		if err := app.handle(ev); err != nil {
+			log.Logger.Warn("failed to handle SUBMIT app event",
+				zap.Error(err))
+		}
+	case states.Accepted:
+		ev := NewRunApplicationEvent(app.GetApplicationId())
+		if err := app.handle(ev); err != nil {
+			log.Logger.Warn("failed to handle RUN app event",
+				zap.Error(err))
+		}
+	case states.Running:
+		if len(app.GetNewTasks()) > 0 {
+			for _, task := range app.getTasks(events.States().Task.New) {
```

Comment:
ApplicationInfo#getTasks will go through taskMap without readLock, may conflict when another thread is updating taskMap. I think we can query new tasks once like this:
```
newTasks := app.GetNewTasks()
if len(newTasks) > 0 {
   for _, task := range newTasks {
   ...
   }
}
```

Reply:
yeah, that's correct. let me fix that. thx